compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -332,15 +332,11 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
     pub(crate) fn smart_resolve_partial_mod_path_errors(
         &mut self,
         prefix_path: &[Segment],
-        path: &[Segment],
+        following_seg: Option<&Segment>,
```

Comment:
Looking at this I'm thinking: The issue we are trying to address by passing only the following segment is to not suggest `Vec::Vec` when we encounter `vec::Vec`. I am thinking a more reliable algorithm could be to take all path segments, and from right to left look for a possible importable type. I'm thinking that if you have `foo::bar::baz::Vec::new()`, and `bar` can't be found, then it should try and suggest `std::vec::Vec::new()`. For that to work we would have to keep the `path: &[Segment]` arg. This is a slightly different case than `my_core::mem::drop`, because we can find `mem` immediately after the unavailable `my_core`.

Does this make sense to you? Would you be interested in trying that approach?

Reply:
This scenario is covered by our code here:
https://github.com/rust-lang/rust/blob/85bf07972a1041b9e25393b803d0e006bec3eaaf/compiler/rustc_resolve/src/late/diagnostics.rs#L528-L536


Usually, we will find some import candidates from the right-most ident:
```rust
let ident = path.last().unwrap().ident; 
```

I will add a test case for it.

Reply:
Added testcase:
https://github.com/rust-lang/rust/blob/6f53e61887b5e918a69915e8f5a721c1f07b9874/tests/ui/resolve/112590-2.stderr#L1-L14